daemon/logger/adapter.go

```diff
@@ -0,0 +1,115 @@
+package logger
+
+import (
+	"encoding/json"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/docker/docker/pkg/plugingetter"
+	"github.com/pkg/errors"
+)
+
+// pluginAdapter takes a plugin and implements the Logger interface for logger
+// instances
+type pluginAdapter struct {
+	driverName   string
+	id           string
+	plugin       logPlugin
+	fifoPath     string
+	stream       io.WriteCloser
+	capabilities Capability
+	mu           sync.Mutex // synchronize access to the log stream
+	logInfo      Info
+
+	// TODO(cpuguy83): not JSON, how about protobuf?
+	// Just using json for quick and dirty implementation
+	// here we are just caching the encoder for writing to the fifo
+	enc *json.Encoder
```

Comment:
FWIW gogo/proto (which I think is already used elsewhere in the code base) includes a couple of simple framing protocols:
- https://github.com/gogo/protobuf/blob/master/io/varint.go
- https://github.com/gogo/protobuf/blob/master/io/uint32.go

So you wouldn't necessarily have to invent one from whole cloth.

I think the uint32 one is similar to how gRPC frames (except gRPC has a byte indicating compression before the 4 byte length).